app/src/main/java/com/google/android/sambadocumentsprovider/provider/SambaDocumentsProvider.java

```diff
@@ -359,11 +399,89 @@ private Cursor buildErrorCursor(String[] projection, @StringRes int resId) {
         case Document.COLUMN_LAST_MODIFIED:
           row[i] = metadata.getLastModified();
           break;
+        case Document.COLUMN_ICON:
+          row[i] = metadata.getIconResourceId();
+          break;
       }
     }
     return row;
   }
 
+  private Object[] getCursorRowForServer(
+          String[] projection,
+          SmbServer server) {
+    Object[] row = new Object[projection.length];
+
+    for (int i = 0; i < projection.length; ++i) {
+      switch (projection[i]) {
+        case Document.COLUMN_DOCUMENT_ID:
+          row[i] = server.getUnresolvedUri();
+          break;
+        case Document.COLUMN_DISPLAY_NAME:
+          row[i] = server.getDisplayName();
+          break;
+        case Document.COLUMN_FLAGS:
+          row[i] = 0;
+          break;
+        case Document.COLUMN_MIME_TYPE:
+          row[i] = Document.MIME_TYPE_DIR;
+          break;
+        case Document.COLUMN_SIZE:
+        case Document.COLUMN_LAST_MODIFIED:
+          row[i] = null;
+          break;
+        case Document.COLUMN_ICON:
+          row[i] = R.drawable.ic_server;
+          break;
+      }
+    }
+
+    return row;
+  }
+
+  private Object[] getCursorRowForBrowsingRoot(String[] projection) {
+    return getCursorRowForServer(projection, browsingRoot);
+  }
+
+  private Cursor getFilesSharesCursor(String[] projection) {
+    final DocumentCursor cursor = new DocumentCursor(projection);
+
+    final Uri uri = toUri(NetworkBrowser.SMB_BROWSING_URI.toString());
+
+    if (mBrowsingStorage == null) {
+      AsyncTask serversTask = mNetworkBrowser.getServersAsync(
+              new OnTaskFinishedCallback<List<SmbServer>>() {
```

Comment:
Save this into a private field to avoid creating an object every time you call this method.